### PR TITLE
Smart Laser Gun

### PR DESCRIPTION
**Pull request by Shadowmech88**
Adds smart laser gun, which can take pAIs by default.
Adds smart laser gun crate to Cargo, for twice the price of a standard weapons crate. It contains four smart laser guns.

Also adds framework for any gun to be usable, on a basic level, by pAI if the can_take_pAI var is changed to TRUE, opening up adminbus options.

:cl:
 * rscadd: Added smart laser guns, which can hold pAIs and be operated by them. Smart laser guns are orderable from Cargo.

### CODE REVIEW

**Comment by Kammerjunk** [comment]
>which can take pAIs

Got my vote. Nice to get some pAI love

**Comment by JustSumBody** [comment]
>orderable from cargo

Oh
NO

**Comment by Shadowmech88** [comment]
I was originally going to put it into the protolathe, but @Sprok0 advised me that Cargo would be better.
At least this way Cargo, the only source of these guns, and Science, the only renewable source of pAIs, will have to collaborate in order for them to be stronger than a normal laser gun.

I'm not particularly attached to one or the other, so I'll change it back if enough people want it to be.

**Comment by jakmak3** [comment]
Why not add them to both cargo order, protolathe, and hell why not add them to the vault locker in armory/replace the xcom lazer in armory

**Comment by aceedex** [comment]
This is p. neat but, I'm wondering a few things which might be a little stupid to ask.
Smart guns can only be fired by the pAI when held in one hand right? can the owner also shoot the gun if there's a pAI inside? do shots fired by the pAI share the same attack delay as the shots fired by the owner?

whatever the answer is though, this is real neat

**Comment by kilozombie** [comment]
Holy christ playing a pAI just got twenty times more interesting.
Can't wait for more smart things, thank you

**Comment by jakmak3** [comment]
Also while we are at it lets brainstorm things that PAI's could do. Now what if PAI chem inject stuff was buffed to match that of borers and instead of being able to have all the chems in a single module you would have to buy each individual chemical in its own module Like having to download tricord or tramadol or coffee

**Comment by kilozombie** [comment]
pAIs shouldn't be mobile chem dispensers, borers are already that and they're pretty lame at times. pAI insertion into items is THE FUTURE.

**Comment by jakmak3** [comment]
hmmmm borers being designed around to work with the body of its host and ignore all outside objects and technology excluding foreign chems and the PAI designed to have minimal interaction with the physiology of its owner and being intended to augment their owners machinery and equipment. I like that idea

**Comment by kilozombie** [comment]
Yeah, that makes sense to me too. pAIs shouldn't match up to borers, they should fulfill a different purpose. The main things the two share is that they help you in some way and they're a friend to talk to at times.

**Comment by jakmak3** [comment]
exactly. Like what if PAI's were capable of tapping into the camrea net OR being able to get the AI view kinda deal of the area near the PAI cards location and also being able to send/recieve messages over any radio freq that was in the owners radio encryption key. Maybe if they could also provide a detailed HUD of the owners health in extreme detail (damage to each limb, alerts any broken bones/bleeding/internal bleeding). What if being able to interact any machines in view of the PAI card (excluding doors/atmos components/engineering machinery)

**Comment by Shadowmech88** [comment]
@aceedex
>Smart guns can only be fired by the pAI when held in one hand right?

The pAI has full control, the master isn't a factor. The pAI can even fire the gun while it's on the ground.
>can the owner also shoot the gun if there's a pAI inside?

Yes.

>do shots fired by the pAI share the same attack delay as the shots fired by the owner?

The gun itself has a fire delay built in, so they can't work together to double-fire, if that's what you mean.
The pAI itself doesn't have an attack delay by default, so it's capable of firing the gun *slightly* faster on its own than a human can fire a gun. I don't think it'll be a problem.

**Comment by kol1th** [comment]
How does it work log-wise?

**Comment by Ephx1** [comment]
I envisioned the smart gun from titanfall based on the title and panicked.
Glad to see pAIs are getting something, and this new system isn't just being tossed to the side.

**Comment by aceedex** [comment]
@Shadowmech88 
It's everything I wanted and more. 
you're alright.
But does that mean you can fire it from a, say, backpack?

**Comment by Sprok0** [comment]
Can it fire from inside a recharger? Can I make smart, rapid fire laser turrets?

**Comment by kilozombie** [comment]
What if you put it in a turret?

**Comment by Shadowmech88** [comment]
@aceedex 
Currently the pAI can fire it from literally anywhere.

I suppose I could restrict it to just turfs and mobs, so it can only fire while on the ground or in someone's hand/pocket.

**Comment by aceedex** [comment]
nah, I don't think it's an issue. leave it as it is.

**Comment by Blithering** [comment]
Anyone else getting a Rogue Trooper vibe from this?

**Comment by Shadowmech88** [comment]
If it becomes a problem in the future, it can always be changed then.

**Comment by kol1th** [comment]
@Shadowmech88 
kol1th commented 27 minutes ago
How does it work log-wise?

**Comment by JustSumBody** [comment]
Actually, that's not so bad then.

**Comment by Shadowmech88** [comment]
The `Fire()` proc logs shots using the `user` var, to which the pAI is passed as an argument.
So the logs should function normally, reading something along the lines of `pAIguy (pAIguy.ckey) fired gun at targetguy (targetguy.ckey)`.

File: code/datums/supplypacks.dm [hunk 1]
```diff
@@ -678,6 +678,18 @@ var/list/all_supply_groups = list("Supplies","Clothing","Security","Hospitality"
...
 
+/datum/supply_packs/smartlaser
+	name = "Smart laser gun crate"
+	contains = list(/obj/item/weapon/gun/energy/laser/smart,
+					/obj/item/weapon/gun/energy/laser/smart,
+					/obj/item/weapon/gun/energy/laser/smart,
+					/obj/item/weapon/gun/energy/laser/smart)
+	cost = 60
+	containertype = /obj/structure/closet/crate/secure/weapon
+	containername = "Smart laser gun crate"
```
Comment (PJB3005) [inline code/datums/supplypacks.dm]: Remove capitalization here.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: code/datums/supplypacks.dm (hunk 1 above)
Comment (Shadowmech88) [reply code/datums/supplypacks.dm]: Should I do the same for every `containername` in the file?

Comment (PJB3005) [reply code/datums/supplypacks.dm]: Nah don't bother.

Comment (Shadowmech88) [reply code/datums/supplypacks.dm]: But then if I fix this it's going to be inconsistent.

Comment (PJB3005) [reply code/datums/supplypacks.dm]: I've told multiple people to do this already, see inflatables.

The thing is, if they're capitalized they'll be proper nouns, so it'll be `That's Smart laser gun crate` vs `That's a smart laser gun crate`.

**Comment by Exxion** [comment]
OK for most things I'd say leave being able to fire from anywhere as-is, but please for the love of god don't allow firing from within pipes or disposals

**Comment by Exxion** [comment]
(For now just hardcode that. I plan on making a system in the future to handle that kind of thing properly, as well as some other stuff)

**Comment by Shadowmech88** [comment]
Do you happen to know offhand of a specific way to check whether something is inside a pipe or disposal pipe?

**Comment by Exxion** [comment]
`get_holder_of_type()` or `get_holder_at_turf_level()`, checking if you're in either kind of pipe

**Comment by Shadowmech88** [comment]
* pAIs are now incapable of firing their gun if they're inside a pipe or disposal pipe

**Review by PJB3005 — APPROVED** [review]

**Pull request merged**